### PR TITLE
Created a nullpoint finder file which finds the nullpoints of a vector space

### PR DESCRIPTION
**Pull request by haman80**
…r space using a trilinear approximation.

- [x] I have added a changelog entry for this pull request.

- [x] If adding new functionality, I have added tests and
      docstrings.

- [x] I have fixed any newly failing tests.

### CODE REVIEW

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by pheuer** [comment]
@rocco8773 

This is great functionality! I do agree that it should be added as methods on `CartesianGrid`, which duplicates most of the functionality in `vector_space`. 

'trilinear_approx' appears to basically be an interpolator (?), so that would also fit nicely as a method of `CartesianGrid`, especially if it was reformatted to take multiple cells in a `pos` array [N, 3] like the other interpolators?

I'm not immediately sure if the best way to implement the null point finder would be 1) as a method of `CartesianGrid` or 2) as a function that takes a `CartesianGrid` and the names of the three vector components as arguments. 

Finally, I found this paper while I was trying to understand what this algorithm did: https://aip.scitation.org/doi/10.1063/1.2756751 
If this algorithm is based of of this paper (or another) we should add a citation in the "Notes" field.

**Comment by namurphy** [comment]
Thank you, @rocco8773 and @pheuer!  My current thinking (which I mentioned to @rocco8773 earlier) is to have the core functionality housed outside of `CartesianGrid`, but to have a method within `CartesianGrid` that calls it (or alternatively a separate function that accepts a `CartesianGrid` as @pheuer brought up).  That would help with separation of responsibilities, and would also reduce the risk of `CartesianGrid` becoming monolithic. I'd prefer to avoid coupling the core functionality too tightly with `CartesianGrid` since there may be people with their own data structures and formats who would want to use this functionality.  I've also been wondering if this functionality could eventually be ported upstream to SciPy, since I can think of a several use cases outside of plasma science.  With that said, I do very much agree that we would ultimately want to encourage people to interface with this functionality via `CartesianGrid` objects.

And of course many thanks to @haman80 for working on this!  I've been looking forward to this functionality for years.

This addresses #764, and is related to #47 and #449.

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

File: plasmapy/analysis/nullpoint.py [hunk 1]
```diff
@@ -0,0 +1,1438 @@
...
+            if p.isEqual(elem):
+                return True
+        return False
+
+    nullpoints = []
+    for i in range(len(vspace[0][0]) - 1):
+        for j in range(len(vspace[0][0][0]) - 1):
+            for k in range(len(vspace[0][0][0][0]) - 1):
+                if reduction(vspace, [i, j, k]):
+                    if trillinear_analysis(vspace, [i, j, k]):
```
Comment (namurphy) [inline plasmapy/analysis/nullpoint.py:1432]:
```suggestion
                    if trilinear_analysis(vspace, [i, j, k]):
```

File: plasmapy/analysis/tests/test_nullpoint.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            [
+                [-5.5, 0, 2, -1, 0, 0, 0, 0],
+                [-22, 3, 0, 1, 0, 0, 0, 0],
+                [-5.96833648, 0.08695652, 1, 0, 0, 0, 0],
+            ],
+        )
+    ]
+    test_trilinear_coeff_cal_expections = []
+
+    test_trilinear_coeff_cal_warnings = []
```
Comment (namurphy) [inline plasmapy/analysis/tests/test_nullpoint.py:58]: Were there still tests that we wanted to add about exceptions and warnings?  If not, we could delete these, at least for the time being.

**Review by namurphy — COMMENTED** [review]
Thank you again for working on this!  I'm really liking how it's coming out and looking forward to getting this merged.  I can tell that you put a lot of thought and effort into both the high-level design and the low-level implementation details.

Most of my suggestions are pretty minor, like typo fixes that should be pretty straightforward to implement.  A bunch of the docstring suggestions are for consistency with other parts of the package more than anything else.  (Unfortunately I haven't had much luck finding docstring autoformatters.)  

My inclination would be to save most of my more in-depth comments for future focused pull requests that we could talk about on Thursday.  For example, it'd be worth thinking about using more array operations (like, instead of using separate variables like `a`, `b`, and `c`, we could use `np.array([a, b, c])`) but that would be easier to do (and quicker to review) in a follow-up pull request.  

One other suggestion would be to add docstrings to the tests as well.  Usually my test docstrings are pretty short, like a sentence or two that describe the purpose of the test, what it's checking, and (infrequently) some additional context to help others understand why the test is being written.

Also, apologies for taking longer to review this than expected!  Apparently I have difficulty focusing during global pandemics sometimes...

File: plasmapy/analysis/nullpoint.py [hunk 3]
```diff
@@ -0,0 +1,1438 @@
...
+    passY = False
+    passZ = False
+    # Check X-Component
+    sign = np.sign(u[cell[0], cell[1], cell[2]])
+    for point in corners:
+        if (
+            u[point[0]][point[1]][point[2]] == 0
+            or np.sign(u[point[0]][point[1]][point[2]]) != sign
+        ):
+            passX = True
```
[the remark this answers is not in the document]
Comment (haman80) [reply plasmapy/analysis/nullpoint.py:491]: I integrated all of the condition checks into a single for loop, so code repetition is avoided. 
I don't believe there is a need for an additional function anymore.

File: plasmapy/analysis/nullpoint.py [hunk 4]
```diff
@@ -0,0 +1,1438 @@
...
+
+    # Checking sign of third component for Bx=By=0
+    if (
+        np.sign(
+            tlApprox(BxByEndpoints[0][0], BxByEndpoints[0][1], BxByEndpoints[0][2])[2]
+        )
+        * np.sign(
+            tlApprox(BxByEndpoints[1][0], BxByEndpoints[1][1], BxByEndpoints[1][2])[2]
+        )
+        > 0
```
[the remark this answers is not in the document]
Comment (haman80) [reply plasmapy/analysis/nullpoint.py:1054]: I created a helper function, so code repetition is reduced.

File: plasmapy/analysis/nullpoint.py [hunk 5]
```diff
@@ -0,0 +1,1357 @@
...
+]
+
+import numpy as np
+import warnings
+
+# Declare Constants & global variables
+_ATOL = 1e-10
+_MAX_DIVIDE = 10
+global _divide
+_divide = 0
```
Comment (namurphy) [inline plasmapy/analysis/nullpoint.py:17]: I'm wondering if it'd be helpful to increase the specificity of the name...perhaps something like `_recursion_level`?

Comment (haman80) [reply plasmapy/analysis/nullpoint.py:17]: I agree. I changed the name of the variable as such.

File: plasmapy/analysis/nullpoint.py [hunk 6]
```diff
@@ -0,0 +1,1357 @@
...
+    This function does not raise any exceptions.
+
+    Warns
+    -----
+    This function does not raise any warnings.
+
+
+    Notes
+    -----
+    N/A
```
Comment (namurphy) [inline plasmapy/analysis/nullpoint.py:331]:
```suggestion
```

**Review by namurphy — COMMENTED** [review]
Thank you again for doing this!

**Review by namurphy — APPROVED** [review]
This looks like a good point to merge it, and follow up with some smaller, focused pull requests.

I'll follow up on this by adding some issues for the code review comments that we decided to postpone (such as improving test coverage, compatibility with `CartesianGrid`, etc.)

@haman80 — thank you so much for doing this!  As I told you before, this is functionality that I've been wanting to have in PlasmaPy for years.

**Pull request merged**